docs/codeql/codeql-language-guides/analyzing-data-flow-in-cpp-new.rst

```diff
@@ -0,0 +1,433 @@
+
+.. code-block:: ql
+
+   nodeFrom.asParameter() = source and
+   nodeTo.asExpr() = sink and
+   TaintTracking::localTaint(nodeFrom, nodeTo)
+
+Examples
+~~~~~~~~
+
+The following query finds the filename passed to ``fopen``.
+
+.. code-block:: ql
+
+   import cpp
+
+   from Function fopen, FunctionCall fc
+   where
+     fopen.hasQualifiedName("fopen") and
+     fc.getTarget() = fopen
+   select fc.getArgument(0)
+
+Unfortunately, this will only give the expression in the argument, not the values which could be passed to it. So we use local data flow to find all expressions that flow into the argument, where we use ``asIndirectExpr(1)`` as we are interested in the value of the string passed to `fopen`, not the pointer pointing to it:
+
+.. code-block:: ql
+
+   import cpp
+   import semmle.code.cpp.dataflow.new.DataFlow
+
+   from Function fopen, FunctionCall fc, Expr src, DataFlow::Node source, DataFlow::Node sink
+   where
+     fopen.hasQualifiedName("fopen") and
+     fc.getTarget() = fopen and
+     source.asIndirectExpr(1) = src and
+     sink.asIndirectExpr(1) = fc.getArgument(0) and
+     DataFlow::localFlow(source, sink)
+   select src
+
+Then we can vary the source and, for example, use the parameter of a function. The following query finds where a parameter is used when opening a file:
+
+.. code-block:: ql
+
+   import cpp
+   import semmle.code.cpp.dataflow.new.DataFlow
+
+   from Function fopen, FunctionCall fc, Parameter p, DataFlow::Node source, DataFlow::Node sink
+   where
+     fopen.hasQualifiedName("fopen") and
+     fc.getTarget() = fopen and
+     source.asParameter(1) = p and
+     sink.asIndirectExpr(1) = fc.getArgument(0) and
+     DataFlow::localFlow(source, sink)
+   select p
+
+The following example finds calls to formatting functions where the format string is not hard-coded.
+
+.. code-block:: ql
+
+   import semmle.code.cpp.dataflow.new.DataFlow
+   import semmle.code.cpp.commons.Printf
+
+   from FormattingFunction format, FunctionCall call, Expr formatString, DataFlow::Node sink
+   where
+     call.getTarget() = format and
+     call.getArgument(format.getFormatParameterIndex()) = formatString and
+     sink.asIndirectExpr(1) = formatString and
+     not exists(DataFlow::Node source |
+       DataFlow::localFlow(source, sink) and
+       source.asIndirectExpr(1) instanceof StringLiteral
+     )
+   select call, "Argument to " + format.getQualifiedName() + " isn't hard-coded."
+
+Exercises
+~~~~~~~~~
+
+Exercise 1: Write a query that finds all hard-coded strings used to create a ``host_ent`` via ``gethostbyname``, using local data flow. (`Answer <#exercise-1>`__)
+
+Global data flow
+----------------
+
+Global data flow tracks data flow throughout the entire program, and is therefore more powerful than local data flow. However, global data flow is less precise than local data flow, and the analysis typically requires significantly more time and memory to perform.
```

Comment:
Another real disadvantage is that the QL to do global flow is more complex / less concise / longer.